### ISSUE TITLE
New refresh token is not used when refreshing session

### ISSUE THREAD

**Issue by carabj** [oauth2-proxy/oauth2-proxy#1187]
According to OAuth2 specification RFC6749,  the Authorization Server can (optionally) return a new refresh token when a refresh token is used to request a new access token. If done, the old refresh token will no longer be valid.

https://tools.ietf.org/html/rfc6749#section-1.5

## Expected Behavior

If the Authorization Server (OIDC Provider) returns a new refresh token when using a refresh token to request a new access token, the new refresh token should be stored in the session and replace the previous refresh token.

## Current Behavior

Only the first refresh token is used on subsequent refresh attempts.

## Possible Solution

If provider return refresh token during refresh, store it and replace the previous one.

## Context

Our Auhorization Server issues new refresh token on every refresh, thus we cannot extend the end user session more than one time.

## Your Environment

- Version used: 7.1.1 with alpha configuration

**Comment by carabj** [comment]
I can add that I'm using the OIDC provider in oauth2-proxy.

**Comment by carabj** [comment]
Hmm, while checking the code of oauth2-proxy and the "oidc.go" provider, it looks like it should use the newly returned refresh token...

**Comment by JoelSpeed** [comment]
This definitely was working in the past, though I haven't tried it recently, so it's possible we could have broken it. Though I would have expected the tests to start failing at that point which I don't think they have 

Have you definitely got a refresh token being created in the session? (You should be able to tell from the logs when it creates a session)

**Comment by carabj** [comment]
Hi Joel,

In the logs I see there is a refresh token when session is created (refresh=true).

It also states refresh = true after first refresh (which is successful).

But as stated, the second time it refresh our authorization server states that the used refresh token is invalid. That’s why I suspected oauth2-proxy used the first refresh token. But I have proof of that..

I could mention that we don’t get an ID Token from our Authorization Server when refreshing the access token. Only new access token and a new refresh token.

One strange thing in the log, is that after the first refresh, it states that the token (or session) expires in just a few minutes. When session was created it correctly states expires in 30 minutes (for my test, TTL for access token is 30 minutes).

I will do some more testing tomorrow. Is it possible to read the content of the stored session? Currently using redis.

Br
Calle

**Comment by JoelSpeed** [comment]
Because of the differences between providers, this feels like this is something that will be pretty hard to reproduce. We have a Dex example environment which we have used for testing this kind of problem in the past, have you had a play with that environment at all? If you can try that and come up with a config that makes that environment break then it's easier for someone like me to help diagnose.

I've definitely seen this working before but I no longer run production systems using OAuth2 Proxy so I haven't checked the latest builds.

The alternative solution here is to build a custom binary with extra log lines in it to inspect the data during the refresh

The data in redis will be encrypted so if you can inspect it, it will just be garbage unless you can work out how to decrypt and decode it, I've never tried doing this myself, I always inspect it in memory with a debugger or with log lines

**Comment by jeffmccune** [comment]
I think I'm running into this too, using `dexidp/dex:v2.28.1` and `quay.io/oauth2-proxy/oauth2-proxy:v7.1.3` with Redis session storage.

I set Dex to expire id tokens after 10m.  Configured oauth2-proxy with `--cookie-refresh=9m`.  I load the kubernetes dashboard, which does a bunch of AJAX.  In the oauth2-proxy logs with `--auth-logging` I see 111 attempts to refresh:

```
grep "Refreshing .* old session cookie for Session" error.txt | wc -l
111
```

The `Session{}` has refresh_token:true token:true id_token:true

In the oauth2-proxy logs around the 10m mark I see:

```
[2021/06/22 00:35:50] [stored_session.go:76] Error loading cookied session: error refreshing access token for session (Session{REDACTED}): error refreshing access token: unable to redeem refresh token: failed to get token: oauth2: cannot fetch token: 500 Internal Server Error
Response: {"error":"server_error"}, removing session
```

At the same time in the Dex logs I see:

```
time="2021-06-22T00:35:50Z" level=error msg="failed to update refresh token: refresh token claimed twice"
time="2021-06-22T00:35:50Z" level=error msg="failed to update refresh token: refresh token claimed twice"
time="2021-06-22T00:35:50Z" level=error msg="failed to update refresh token: refresh token claimed twice"
time="2021-06-22T00:35:50Z" level=error msg="failed to update refresh token: refresh token claimed twice"
time="2021-06-22T00:35:50Z" level=error msg="failed to update refresh token: refresh token claimed twice"
time="2021-06-22T00:35:50Z" level=error msg="failed to update offline session: maximum timeout reached while retrying a conflicted request"
time="2021-06-22T00:35:50Z" level=error msg="failed to update refresh token: refresh token claimed twice"
```

Time permitting I'll dig into this, I saw a couple of related issues.

**Comment by JoelSpeed** [comment]
Ok, this definitely sounds like we are redeeming the refresh token, but not persisting it for some reason. The provider logic is where we need to look, `RefreshSessionIfNeeded`

**Comment by jeffmccune** [comment]
Thanks for your hard work on the refresh and locking @JoelSpeed @NickMeves and @Bibob7.  I'm using your refresh and session update functionality with Istio [External Authorization](https://docs.google.com/document/d/1V4mCQCw7mlGp0zSQQXYoBdbKMDnkPOjeyUb85U07iSI/edit#).  I'm hoping to handle the use case where a chatty AJAX app like the k8s dashboard can stay in the foreground and oauth2-proxy keeps the id token fresh inside the service mesh.

I started looking into this by building an image of `v7.1.3-32-g16a9893` which includes #1086 and #1063.  Leaving the k8s dashboard in a foreground window results in a dead session around the 10 min id token expiration with the following logs.

Let me know if you'd like me to report back somewhere else, otherwise I'll dive deeper and follow up with what I find out here.

```
[2021/06/23 03:46:35] [stored_session.go:77] Error loading cookied session: failed to load the session state with the ticket: error loading redis session: context canceled, removing session
[2021/06/23 03:46:35] [stored_session.go:80] Error removing session: error clearing the session from redis: context canceled
[2021/06/23 03:47:16] [stored_session.go:77] Error loading cookied session: http: named cookie not present, removing session
```

**Comment by jeffmccune** [comment]
I think I see it.  oauth2-proxy doesn't seem to react to the id token expiration time.  Is this expected and --cookie-expiration should always be shorter than id token expiration?  I set id expiry to 2min in Dex and got a fresh id token at `time="2021-06-23T04:37:54Z"` at the 2 minute mark the browser gets Jwt expired errors.

oauth2-proxy doesn't try to refresh until 5 min later, which coincidentally aligns with the `--cookie-refresh=5m` setting I used with the 2m id token expiry.

```
[2021/06/23 04:45:32] [stored_session.go:119] Refreshing session - User: CgpqZWZmbWNjdW5lEgZnaXRodWI; SessionAge: 7m37.734415012s
```

I think this should work because refresh tokens have an infinite TTL in Dex, but doesn't seem to because of the redis locking?  Or maybe I'm off track here.

```
time="2021-06-23T04:45:33Z" level=error msg="failed to update refresh token: refresh token claimed twice"
```

**Comment by NickMeves** [comment]
> I think I see it. oauth2-proxy doesn't seem to react to the id token expiration time. Is this expected and --cookie-expiration should always be shorter than id token expiration? I set id expiry to 2min in Dex and got a fresh id token at `time="2021-06-23T04:37:54Z"` at the 2 minute mark the browser gets Jwt expired errors.
> 
> oauth2-proxy doesn't try to refresh until 5 min later, which coincidentally aligns with the `--cookie-refresh=5m` setting I used with the 2m id token expiry.
> 
> ```
> [2021/06/23 04:45:32] [stored_session.go:119] Refreshing session - User: CgpqZWZmbWNjdW5lEgZnaXRodWI; SessionAge: 7m37.734415012s
> ```
> 
> I think this should work because refresh tokens have an infinite TTL in Dex, but doesn't seem to because of the redis locking? Or maybe I'm off track here.
> 
> ```
> time="2021-06-23T04:45:33Z" level=error msg="failed to update refresh token: refresh token claimed twice"
> ```

This PR I just merged will allow refreshing before an IDToken is expired: https://github.com/oauth2-proxy/oauth2-proxy/pull/1086

Until we have a release with that fix, our recommendation is to manually set your cookie-refresh to the same time as your IDToken TTL.

**Comment by jeffmccune** [comment]
> This PR I just merged will allow refreshing before an IDToken is expired: #1086

@NickMeves My two comments above your's were run with v7.1.3-32-g16a9893 which includes #1086 and #1063.  Are there any additional changes I need to make which will allow refreshing before an IDToken expires?

**Comment by NickMeves** [comment]
Do you have the `cookie-refresh` setting set to a length of time less than the TTL of your IDToken?

**Comment by jeffmccune** [comment]
Both were set to 10 min exactly when I tested the comment directly above your last.

I'm re-testing now with Dex `idTokens: "10m"` and oauth2-proxy `--cookie-refresh=9m`

**Comment by jeffmccune** [comment]
@NickMeves `cookie-refresh` 9m and `idTokens: "10m"` results in the following log messages, but the dashboard remains operational beyond 10 minutes (I'm at the 15 minute mark now) and I don't see the JWT expired errors this time.

```
[2021/06/29 15:24:15] [stored_session.go:119] Refreshing session - User: XXXXXXXXXXXXXXXXXXXXXXXXXXX; SessionAge: 9m3.943664565s
[2021/06/29 15:24:15] [stored_session.go:119] Refreshing session - User: XXXXXXXXXXXXXXXXXXXXXXXXXXX; SessionAge: 9m3.943664565s
[2021/06/29 15:24:16] [stored_session.go:124] Unable to refresh session: error refreshing tokens: unable to redeem refresh token: failed to get token: oauth2: cannot fetch token: 500 Internal Server Error
Response: {"error":"server_error"}
```

And in Dex:

```
time="2021-06-29T15:24:16Z" level=error msg="failed to update refresh token: refresh token claimed twice"
```

**Comment by NickMeves** [comment]
I wonder if you are bumping into these issues @Bibob7 is trying to work through here: https://github.com/oauth2-proxy/oauth2-proxy/pull/1063

In dashboard / single-page apps where many requests are in parallel; during the refresh they race and make multiple requests.

If your IdP doesn't allow that, you get into trouble.

Hence the work to add refresh lock support.

That seems to be what happened -- that first request refreshed, the second one was sent with a stale accesstoken and was denied?

**Comment by jeffmccune** [comment]
I think so, yes.  I have merged in #1063 in this custom build, perhaps it's
not yet complete as of commit 16a9893.

Dex refresh tokens are single use only.

**Comment by xXluki98Xx** [comment]
Hi,
I am not sure if its the same topic, but I run into a similar error on oauth2-proxy+cookie+zitadel.
With redis it seems to work, but there is another error i cant get at the moment.

I am using the 7.6.0 version of oauth2-proxy, but with the bitnami helmchart version 5.3.3.
